libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt

```diff
@@ -250,22 +265,305 @@ internal class NativeLoginManager(
+     * "email" or "sms"
+     */
+    private data class OtpRequestBody(
+        val recaptcha: String?,
+        val recaptchaevent: OtpRequestBodyReCaptchaEvent?,
+        val username: String,
+        val verificationMethod: String
+    ) {
+        // TODO: Inquire regarding modern JSON serialization in MSDK. ECJ20240317
+        fun toJson() = JSONObject().apply {
+            put("recaptcha", recaptcha)
+            put("recaptchaevent", recaptchaevent)
+            put("username", username)
+            put("verificationmethod", verificationMethod)
+        }
+    }
+
+    /**
+     * A data class for the OTP request body's reCAPTCHA event parameter.
+     * @param token The reCAPTCHA token provided by the reCAPTCHA Android SDK.
+     * This is used only with reCAPTCHA Enterprise
+     * @param siteKey The Google Cloud project reCAPTCHA Key's "Id" as shown in
+     * Google Cloud Console under "Products & Solutions", "Security" and
+     * "reCAPTCHA Enterprise"
+     * @param expectedAction The user-initiated "Action Name" for the reCAPTCHA
+     * event.  A specific value is not required by Google though it is used in
+     * reCAPTCHA Metrics.  "login" is a recommended value from Google
+     * documentation
+     * @param projectId The Google Cloud project's "Id" as shown in Google Cloud
+     * Console
+     */
+    private data class OtpRequestBodyReCaptchaEvent(
+        val token: String,
+        val siteKey: String,
+        val expectedAction: String = "login",
+        val projectId: String
+    )
+
+    // endregion
+    // region Private Implementation
+
+    /**
+     * Generates a Base64 encoded value by concatenating the provided values
+     * with a colon, which is commonly used in the Headless Identity API request
+     * headers.
+     * @param value1 The left-side value
+     * @param value2 The right-side value
+     * @return The Base64 encoded value
+     */
+    private fun generateColonConcatenatedBase64String(
+        value1: String,
+        value2: String
+    ) = encodeToString(
+        "$value1:$value2".toByteArray(),
+        URL_SAFE or NO_WRAP or NO_PADDING
+    )
+
+    /**
+     * Reacts to a response from the Headless Identity API authorization
+     * endpoint to initiate the token exchange, request a granted access token
+     * and create the user's session.
+     * @param authorizationResponse The result from the Headless Identity API
+     * authorization endpoint
+     * @param codeVerifier The code verifier
+     * @return The native login result
+     */
+    private suspend fun submitAccessTokenRequest(
+        authorizationResponse: RestResponse,
+        codeVerifier: String
+    ): NativeLoginResult {
+        if (authorizationResponse.isSuccess) {
+            val code = authorizationResponse.asJSONObject().get(CODE).toString()
+            val authEndpoint = authorizationResponse.asJSONObject().get(SFDC_COMMUNITY_URL).toString()
+            val tokenRequestBody = createRequestBody(
```

Comment:
Doesn't the Oauth2 exchangeCode method do that?

Reply:
Shouldn't we use HYBRID_AUTH_CODE for the grant type when hybrid auth flow is enabled? See https://github.com/forcedotcom/SalesforceMobileSDK-Android/blob/dev/libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/OAuth2.java#L320

That causes the token end point to return a bunch of cookies that some app (e.g. SApp) can use when "hydrating" their web view.

Reply:
That one is on me.  I missed that in my username/password PR.  

And I believe we aren't using OAuth2's `exchangeCode` because we need to handle/suspend the TokenEndpointResponse (which involves making the user info request) to return success or failure at the right time.  If we don't the app will call finish on the native login view too early and it will just get recreated and sit in the background forever (since we won't have launched he main activity yet).

Reply:
iOS also does not use hybrid_auth_code.  😬

Reply:
@wmathurin and @brandonpage - Do we need a change here? I read Brandon's note above as "works as expected", but wasn't sure.

Reply:
I wrote a crossplatform WI for this.